File: src/main/java/io/github/tanguygab/tabadditions/shared/features/actionbar/ActionBarManager.java
package io.github.tanguygab.tabadditions.shared.features.actionbar;

import io.github.tanguygab.tabadditions.shared.TABAdditions;
import io.github.tanguygab.tabadditions.shared.commands.ActionBarCmd;
import lombok.Getter;
import me.neznamy.tab.shared.config.file.ConfigurationFile;
import me.neznamy.tab.shared.config.file.ConfigurationSection;
import me.neznamy.tab.shared.cpu.CpuManager;
import me.neznamy.tab.shared.cpu.TimedCaughtTask;
import me.neznamy.tab.shared.features.PlaceholderManagerImpl;
import me.neznamy.tab.shared.features.types.*;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.placeholders.conditions.Condition;
import me.neznamy.tab.shared.platform.TabPlayer;
import me.neznamy.tab.shared.util.cache.StringToComponentCache;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public class ActionBarManager extends RefreshableFeature implements UnLoadable, CommandListener, JoinListener {

    private static final StringToComponentCache cache = new StringToComponentCache("ActionBar", 100);

    @Getter private final String featureName = "ActionBar";
    @Getter private final String refreshDisplayName = "&aActionBar&r";
    @Getter private final String command = "/toggleactionbar";

    private final TABAdditions plugin;
    @Getter private final Map<String, ActionBarLine> actionBars = new LinkedHashMap<>();
    private final Map<TabPlayer, String> announcedBars = new HashMap<>();
    private final List<UUID> toggled;
    private final boolean toggleCmd;

    public ActionBarManager() {
        plugin = TABAdditions.getInstance();
        TAB tab = TAB.getInstance();
        tab.getCommand().registerSubCommand(new ActionBarCmd(this));

        ConfigurationFile config = plugin.getConfig();
        toggleCmd = config.getBoolean("actionbars./toggleactionbar",true);
        if (toggleCmd) plugin.getPlatform().registerCommand("toggleactionbar");
        toggled = plugin.loadData("actionbar-off",toggleCmd);

        ConfigurationSection barsConfig = config.getConfigurationSection("actionbars.bars");
        barsConfig.getKeys().forEach(key -> {
            String bar = key.toString();
            ConfigurationSection section = barsConfig.getConfigurationSection(bar);

            String text = section.getString("text");
            String condition = section.getString("condition");

            if (text != null) addUsedPlaceholders(PlaceholderManagerImpl.detectPlaceholders(text));
            actionBars.put(bar,new ActionBarLine(text,condition == null ? null : Condition.getCondition(condition)));
        });

        tab.getCPUManager().getProcessingThread().repeatTask(new TimedCaughtTask(tab.getCPUManager(), ()->{
            for (TabPlayer p : tab.getOnlinePlayers())
                refresh(p,false);
        },featureName,"handling ActionBar"),2000);
    }

    @Override
    public void unload() {
        plugin.unloadData("actionbar-off",toggled,toggleCmd);
    }

    @Override
    public void refresh(@NotNull TabPlayer player, boolean force) {
        String text;
        if (announcedBars.containsKey(player)) {
            text = announcedBars.get(player);
            if (actionBars.containsKey(text))
                text = actionBars.get(text).getText();
        } else {
            ActionBarLine bar = getActionBar(player);
            if (bar == null) return;
            text = bar.getText();
        }
        if (toggled.contains(player.getUniqueId())) return;

        text = plugin.parsePlaceholders(text,player);
        plugin.getPlatform().sendActionbar(player, plugin.toFlatText(cache.get(text)));
    }

    public ActionBarLine getActionBar(TabPlayer player) {
        for (ActionBarLine bar : actionBars.values())
            if (bar.isConditionMet(player))
                return bar;
        return null;
    }

    public void announceBar(TabPlayer player, String actionbar) {
        addUsedPlaceholders(PlaceholderManagerImpl.detectPlaceholders(actionbar));
        announcedBars.put(player,actionbar);
        refresh(player,true);

        CpuManager cpu = TAB.getInstance().getCPUManager();
        cpu.getProcessingThread().executeLater(new TimedCaughtTask(cpu, () -> {
            if (actionbar.equals(announcedBars.get(player))) announcedBars.remove(player);
        }, featureName, "handling ActionBar on join for "+player.getName()),2000);
    }

    @Override
    public void onJoin(TabPlayer player) {
        String property = player.loadPropertyFromConfig(this,"join-actionbar", "").getCurrentRawValue();
        if (property.isEmpty()) return;
        announceBar(player,property);
    }

    @Override
    public boolean onCommand(@NotNull TabPlayer player, String msg) {
        if (msg.equals(command) && plugin.toggleCmd(toggleCmd,player,toggled,plugin.getTranslation().actionBarOn,plugin.getTranslation().actionBarOff)) {
            refresh(player,true);
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/actionbar/ActionBarManager.kt
package io.github.tanguygab.tabadditions.shared.features.actionbar

import io.github.tanguygab.tabadditions.shared.TABAdditions
import io.github.tanguygab.tabadditions.shared.commands.ActionBarCmd
import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.cpu.TimedCaughtTask
import me.neznamy.tab.shared.features.PlaceholderManagerImpl
import me.neznamy.tab.shared.features.types.CommandListener
import me.neznamy.tab.shared.features.types.JoinListener
import me.neznamy.tab.shared.features.types.RefreshableFeature
import me.neznamy.tab.shared.features.types.UnLoadable
import me.neznamy.tab.shared.platform.TabPlayer
import me.neznamy.tab.shared.util.cache.StringToComponentCache
import java.util.UUID

class ActionBarManager : RefreshableFeature(), UnLoadable, CommandListener, JoinListener {
    private val plugin: TABAdditions = TABAdditions.INSTANCE

    override fun getFeatureName() = "ActionBar"
    override fun getRefreshDisplayName() = "&aActionBar&r"
    override fun getCommand() = "/toggleactionbar"

    val actionBars = mutableMapOf<String, ActionBarLine>()
    private val announcedBars = mutableMapOf<TabPlayer, String>()
    private val toggled: MutableList<UUID>
    private val toggleCmd: Boolean

    init {
        val tab = TAB.getInstance()
        tab.command.registerSubCommand(ActionBarCmd(this))

        toggleCmd = plugin.config.getBoolean("actionbars./toggleactionbar", true)
        if (toggleCmd) plugin.platform.registerCommand("toggleactionbar")
        toggled = plugin.loadData("actionbar-off", toggleCmd)

        val barsConfig = plugin.config.getConfigurationSection("actionbars.bars")
        barsConfig.keys.forEach {
            val bar = it.toString()
            val section = barsConfig.getConfigurationSection(bar)

            val text = section.getString("text") ?: ""
            val condition = section.getString("condition")

            if (text.isNotEmpty()) addUsedPlaceholders(PlaceholderManagerImpl.detectPlaceholders(text))
            actionBars[bar] = ActionBarLine(
                text,
                if (condition == null) null
                else tab.placeholderManager.conditionManager.getByNameOrExpression(condition)
            )
        }

        tab.cpuManager.processingThread.repeatTask(TimedCaughtTask(tab.cpuManager, {
            tab.onlinePlayers.forEach { refresh(it, false) }
        }, featureName, "handling ActionBar"), 2000)
    }

    override fun unload() {
        plugin.unloadData("actionbar-off", toggled, toggleCmd)
    }

    override fun refresh(player: TabPlayer, force: Boolean) {
        var text = if (player in announcedBars) {
            val bar = announcedBars[player]!!
            actionBars[bar]?.text ?: bar
        } else getActionBar(player)?.text ?: return

        if (player.uniqueId in toggled) return

        text = plugin.parsePlaceholders(text, player)
        plugin.platform.sendActionbar(player, plugin.toFlatText(cache[text]))
    }

    fun getActionBar(player: TabPlayer) = actionBars.values.find { it.isConditionMet(player) }

    fun announceBar(player: TabPlayer, actionbar: String) {
        addUsedPlaceholders(PlaceholderManagerImpl.detectPlaceholders(actionbar))
        announcedBars[player] = actionbar
        refresh(player, true)

        val cpu = TAB.getInstance().cpuManager
        cpu.processingThread.executeLater(TimedCaughtTask(cpu, {
            if (actionbar == announcedBars[player]) announcedBars.remove(player)
        }, featureName, "handling ActionBar on join for " + player.name), 2000)
    }

    override fun onJoin(player: TabPlayer) {
        val property = player.loadPropertyFromConfig(this, "join-actionbar", "").currentRawValue
        if (property.isEmpty()) return
        announceBar(player, property)
    }

    override fun onCommand(player: TabPlayer, msg: String): Boolean {
        if (msg == command && plugin.toggleCmd(
                toggleCmd,
                player,
                toggled,
                plugin.translation.actionBarOn,
                plugin.translation.actionBarOff
            )
        ) {
            refresh(player, true)
            return true
        }
        return false
    }

    companion object {
        private val cache = StringToComponentCache("ActionBar", 100)
    }
}
